File: 프로그래머스/2/42577. 전화번호 목록/전화번호 목록.js
function solution(phone_book) {
    
    // 방법 1. 전체 phone_book 배열 정렬(문자열 기준) 후 이후 값이 이전 값으로 시작하는지 확인
    
    // 방법 2. 해시 맵 사용 => 우선 이거로 풀어보자.
    const hashMap = new Map();
    for (const number of phone_book) {
        hashMap.set(number, true); // 값이 있는지 없는지 판단하고 싶을 때, 그냥 true 넣어주자.
    }
    
    for (const number of phone_book) {
        for (let i = 0; i < number.length; i++) {
            if (hashMap.get(number.slice(0, i)) === true) return false;
        }
    }
    
    return true;
}

File: 프로그래머스/2/42577. 전화번호 목록/전화번호 목록.java
import java.util.*;

class Solution {
    public boolean solution(String[] phone_book) {
        Arrays.sort(phone_book);
        
        for (int i = 0; i < phone_book.length - 1; i++) {
            String prefix = phone_book[i];
            String next = phone_book[i+1];
            if (next.length() > prefix.length() && next.substring(0, prefix.length()).equals(prefix)) 
                return false;
        }
        
        return true;
    }
}